### PR TITLE
add bottlerocket tools to twoliter binary

### LINKED ISSUE #9: copy tools
### Description

When building the Twoliter bootstrapping container, copy the necessary build tools into place.
In __ we defined a container using the SDK as its base.
In this issue we will add:

- `buildsys`, `pubsys`, etc.
- `rpm2img` and all other sdk-environment scripts.
- Bottlerocket's `Makefile.toml`

The above items should have their git history preserved.

### Definition of Done

When Twoliter builds its container on the fly and copies in the items listed above.

#### Depends on:

- [x] https://github.com/bottlerocket-os/twoliter/issues/8
- [x] https://github.com/bottlerocket-os/twoliter/issues/10

```
id: texec_tools_cpy
epic: twoliter_exec
```

Related to https://github.com/bottlerocket-os/bottlerocket/issues/2669

### LINKED ISSUE #14: twoliter make can build bottlerocket
### Description

Get the `twoliter make` command working for all `Makefile.toml` targets.
This will require mounting the source repo into the Twoliter execution environment.

### Definition of Done

When all of the `cargo make` targets in the Bottlerocket `Makefile.toml` can be executed through `twoliter make`.

#### Depends on:

- [x] https://github.com/bottlerocket-os/twoliter/issues/13
- [x] https://github.com/bottlerocket-os/twoliter/issues/7
- [x] https://github.com/bottlerocket-os/twoliter/issues/9
- [x] https://github.com/bottlerocket-os/twoliter/issues/32
- [x] https://github.com/bottlerocket-os/bottlerocket/issues/3294

```
id: texec_command_works
epic: twoliter_exec
```

Related to https://github.com/bottlerocket-os/bottlerocket/issues/2669

### LINKED ISSUE #32: change build tools install dir
### Description

Currently, the location of `buildsys` and other binaries is fixed in the repo at `tools/bin`.
This will not work with out-of-tree builds where we need to use pre-built build tools.
Make it possible to install and call the build tools binaries from arbitrary locations.

### Definition of Done

A PR is merged that allows the user to decide where `buildsys` and other build tools will be installed and called from.

`id: br_toolsdir`
Related to https://github.com/bottlerocket-os/bottlerocket/issues/2669

### PR DESCRIPTION
**Pull request by webern**
*Issue #, if available:*

Closes #14 
Closes #9 
Closes #32
Partly Addresses #43 (We install it here, but should probably still eliminate it in favor of a lib)

*Description of changes:*

Embed tools such as `buildsys`, and scripts such as `rpm2img` into the Twoliter binary. Use them when building Bottlerocket

*Testing*

- [x] Build Bottlerocket on an x86_64 host
- [x] Build Bottlerocket on an Arm host

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[unstable]
+bindeps = true
+
+[env]
+RUSTC_BOOTSTRAP = "1"
```
Comment (webern) [inline .cargo/config.toml]: Improve this.
See https://github.com/axodotdev/cargo-dist/issues/347#issuecomment-1698141019

File: tools/buildsys/src/builder.rs [hunk 2]
```diff
@@ -121,6 +133,11 @@ impl PackageBuilder {
             args.build_arg(target_env_var, src_env_val);
         }
 
+        // Add the tools directory as a secret since it is not in the context directory.
+        args.extend(tools_secret_args(&PathBuf::from(
```
Comment (webern) [inline tools/buildsys/src/builder.rs:136]: Right now I'm still passing secrets arguments to the docker build command even though they are not needed because Twoliter is providing an image that has the tools FROM scratch.

I like the `FROM scratch` image better than using secrets, just because it feels cleaner. But I need to choose one method and clear away the detritus from the other method.

File: tools/buildsys/src/builder.rs [hunk 3]
```diff
@@ -650,3 +681,30 @@ where
...
+//             if let Some(filename) = filename.to_str() {
+//                 if path.is_file() && !filename.starts_with(".") {
+//                     files.push(path);
+//                 }
+//             }
+//         }
+//     }
+//     Ok(files)
+// }
+
```
Comment (webern) [inline tools/buildsys/src/builder.rs]:
```suggestion
```

File: tools/buildsys/src/gomod.rs [hunk 4]
```diff
@@ -188,8 +189,12 @@ fn docker_go(root_dir: &Path, dg_args: &DockerGoArgs) -> Result<()> {
         &dg_args.command,
     ];
     let arg_string = args.join(" ");
-    let program = root_dir.join("tools/docker-go");
-    println!("program: {}", program.to_string_lossy());
+    // TODO the getenv function returns an error type that is very inconvenient here.
```
Comment (webern) [inline tools/buildsys/src/gomod.rs]: I think I figured this out anyway, clean this up.

File: tools/buildsys/src/gomod.rs [hunk 5]
```diff
@@ -198,7 +203,7 @@ fn docker_go(root_dir: &Path, dg_args: &DockerGoArgs) -> Result<()> {
         .context(error::CommandStartSnafu)?;
 
     let stdout = String::from_utf8_lossy(&output.stdout);
-    println!("{}", &stdout);
+    eprintln!("{}", &stdout);
```
Comment (webern) [inline tools/buildsys/src/gomod.rs]: Revert this.

File: tools/buildsys/src/builder/error.rs [hunk 6]
```diff
@@ -59,6 +59,9 @@ pub(crate) enum Error {
         source: std::io::Error,
     },
 
+    #[snafu(display("{}", source))]
+    GoModError { source: crate::gomod::error::Error },
+
```
Comment (webern) [inline tools/buildsys/src/builder/error.rs]:
```suggestion
```

File: tools/buildsys/src/gomod/error.rs [hunk 7]
```diff
@@ -35,6 +35,11 @@ pub(crate) enum Error {
         source: std::io::Error,
     },
 
+    /// The error type returned by getenv, located currently in main.rs, cannot be handled properly
+    /// because it would be recursive here.
+    #[snafu(display("{}", message))]
+    MainError { message: String },
+
```
Comment (webern) [inline tools/buildsys/src/gomod/error.rs]:
```suggestion
```

File: tools/buildsys/src/gomod/error.rs [hunk 8]
```diff
@@ -55,3 +60,11 @@ pub(crate) enum Error {
...
 
 pub(super) type Result<T> = std::result::Result<T, Error>;
+
+impl From<crate::error::Error> for Error {
+    fn from(value: crate::error::Error) -> Self {
+        Error::MainError {
+            message: value.to_string(),
+        }
+    }
+}
```
Comment (webern) [inline tools/buildsys/src/gomod/error.rs]:
```suggestion
```

File: twoliter/build.rs [hunk 9]
```diff
@@ -10,57 +10,128 @@ Prepare and package embedded tools in a tarball to be included with Twoliter.
...
+    ));
+    println!("Done at {:?}", SystemTime::now());
+}
+
+struct Paths {
+    /// The directory where our scripts, Makefile.toml etc. are located.
+    data_input_dir: PathBuf,
+    /// The directory that we will copy everything to before creating a tarball.
+    prep_dir: PathBuf,
+    /// The path to
```
Comment (webern) [inline twoliter/build.rs]:
```suggestion
    /// The path to a file containing the hash of the `Twoliter.dockerfile` image input.
```

File: twoliter/embedded/Dockerfile [hunk 10]
```diff
@@ -9,12 +9,16 @@
...
+ARG TOOLS_IMAGE
 ARG ARCH
 ARG GOARCH
 
-FROM ${SDK} as sdk
+FROM ${TOOLS_IMAGE} as tools
 FROM --platform=linux/${GOARCH} ${TOOLCHAIN}-${ARCH} as toolchain
 
+FROM ${SDK} as sdk
+COPY --chown=1000:1000 --from=tools /tools/ /tools/
```
Comment (webern) [inline twoliter/embedded/Dockerfile]:
```suggestion
FROM ${TOOLS_IMAGE} as tools

FROM ${SDK} as sdk
COPY --chown=1000:1000 --from=tools /tools/ /tools/

FROM --platform=linux/${GOARCH} ${TOOLCHAIN}-${ARCH} as toolchain
```

File: twoliter/embedded/Makefile.toml [hunk 11]
```diff
@@ -2,6 +2,10 @@
 skip_core_tasks = true
 
 [env]
+# The directory that twoliter has installed tools like the buildsys binary,
+# rpm2img, etc. There can be no reasonable default for this, it must be
+# specified when running cargo make.
```
Comment (webern) [inline twoliter/embedded/Makefile.toml]: Also validate that it is not empty.

```suggestion
# The directory into which Twoliter has installed tools (e.g. buildsys,
# rpm2img, etc.) There can be no reasonable default for this, it must be
# specified when running cargo make.
```

File: twoliter/src/cmd/make.rs [hunk 12]
```diff
@@ -64,6 +68,17 @@ impl Make {
     }
 }
 
+// TODO - remove
```
Comment (webern) [inline twoliter/src/cmd/make.rs]: ?

File: twoliter/src/docker/image.rs [hunk 13]
```diff
@@ -32,6 +38,47 @@ impl ImageUri {
             Some(registry) => format!("{}/{}:{}", registry, self.repo, self.tag),
         }
     }
+
+    /// Uses `docker image inspect` to discover whether the image exists on the local host.
```
Comment (webern) [inline twoliter/src/docker/image.rs]:
```suggestion
    /// Uses `docker inspect` to discover whether the image exists on the local host.
```

File: twoliter/src/tools/mod.rs [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        .execute()
+        .await?;
+    Ok(uri)
+}
+
+/// A sha digest of the canonicalized `project_dir`. We can use this to avoid colliding with other
+/// Twoliter project builds, or multiple checkouts of the same Twoliter project.
+fn instance_token(project_dir: impl AsRef<Path>) -> Result<String> {
+    let path = canonicalize(project_dir)?;
+    // Compute a per-checkout prefix for the tag to avoid collisions.
```
Comment (webern) [inline twoliter/src/tools/mod.rs]:
```suggestion
```

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/03e54a7a9a6f157f339e99d65a811bc50b1d09e1..b6c81c329538bc1259413081d604fd2646419f66

Clean some things up.

File: tools/buildsys/src/builder.rs [hunk 15]
```diff
@@ -83,6 +85,15 @@ lazy_static! {
...
 static DOCKER_BUILD_MAX_ATTEMPTS: NonZeroU16 = nonzero!(10u16);
 
+/// The list of tools.
+const TOOLS: &[&str] = &[
+    "docker-go",
+    "partyplanner",
+    "rpm2img",
+    "rpm2kmodkit",
+    "rpm2migrations",
+];
```
Comment (bcressey) [inline tools/buildsys/src/builder.rs:95]: nit: the buildsys changes might be better as their own commits

File: twoliter/embedded/Dockerfile [hunk 16]
```diff
@@ -13,6 +13,18 @@ ARG ARCH
...
+USER root
+RUN \
+     --mount=type=secret,id=tools-docker-go,mode=755,target=/tmp/tools/docker-go \
+     --mount=type=secret,id=tools-partyplanner,mode=755,target=/tmp/tools/partyplanner \
+     --mount=type=secret,id=tools-rpm2img,mode=755,target=/tmp/tools/rpm2img \
+     --mount=type=secret,id=tools-rpm2kmodkit,mode=755,target=/tmp/tools/rpm2kmodkit \
+     --mount=type=secret,id=tools-rpm2migrations,mode=755,target=/tmp/tools/rpm2migrations \
+     cp -a /tmp/tools/ /tools/ \
+     && chown -R 1000:1000 /tools/ \
+     && chmod 755 /tools/
```
Comment (bcressey) [inline twoliter/embedded/Dockerfile]: nit: are the `chown` and `chmod` necessary?

File: twoliter/embedded/Makefile.toml [hunk 17]
```diff
@@ -233,6 +234,17 @@ fi
 '''
 ] }
 
+# Ensures TWOLITER_TOOLS_DIR is non-zero-length. This variable is unused.
+UNUSED_TWOLITER_TOOLS_DIR_CHECK = { script = [
+'''
+if [ -z "${TWOLITER_TOOLS_DIR}" ];then
+   echo "TWOLITER_TOOLS_DIR must be defined and must be non-zero in length."
+   exit 1
+fi
```
Comment (bcressey) [inline twoliter/embedded/Makefile.toml]: Could we roll this into `tasks.setup` like the architecture check?

File: twoliter/embedded/Makefile.toml [hunk 18]
```diff
@@ -464,19 +461,6 @@ rc=0
...
-if ! docker run --rm \
-   -u $(id -u):$(id -g) \
-   -e CARGO_HOME="/tmp/.cargo" \
-   -v "${CARGO_HOME}":/tmp/.cargo \
-   -v "${BUILDSYS_ROOT_DIR}/sources":/tmp/sources \
-   -v "${BUILDSYS_ROOT_DIR}/tools":/tmp/tools \
-   "${BUILDSYS_SDK_IMAGE}" \
-   cargo clippy \
-  --manifest-path /tmp/tools/Cargo.toml \
-  --locked -- -D warnings --no-deps; then
```
Comment (bcressey) [inline twoliter/embedded/Makefile.toml:476]: The `check-shell` test was previously linting some of the migrated scripts so we should bring that back to this repo at some point.

File: twoliter/embedded/Makefile.toml [hunk 19]
```diff
@@ -939,25 +867,12 @@ dependencies = [
...
-]
-
 [tasks.create-infra]
-dependencies = ["infra-tools"]
 script = [
 '''
 set -e
 
-export PATH="${BUILDSYS_TOOLS_DIR}/bin:${PATH}"
+export PATH="${TWOLITER_TOOLS_DIR}:${PATH}"
```
Comment (bcressey) [inline twoliter/embedded/Makefile.toml]: Are we including `infrasys` also? I'd be OK with just dropping it here, not including it as a tool, and removing the sources later.

File: twoliter/build.rs [hunk 20]
```diff
@@ -11,56 +11,108 @@ use bytes::BufMut;
...
 
-    // This is the filepath to the tarball we will create.
-    let tar_path = out_dir.join("tools.tar.gz");
+    paths.copy_file("Dockerfile");
+    paths.copy_file("Makefile.toml");
+    paths.copy_file("docker-go");
+    paths.copy_file("partyplanner");
+    paths.copy_file("rpm2img");
+    paths.copy_file("rpm2kmodkit");
+    paths.copy_file("rpm2migrations");
```
Comment (bcressey) [inline twoliter/build.rs:34]: Do we need to emit `rerun-if-changed` for the non-Rust binary artifacts?

**Review by bcressey — APPROVED** [review]

File: .github/workflows/rust.yml [hunk 21]
```diff
@@ -18,6 +18,8 @@ on:
 jobs:
   build:
     runs-on: ubuntu-latest
+    env:
+      RUSTC_BOOTSTRAP: "1"
```
Comment (cbgbt) [inline .github/workflows/rust.yml]: I've never seen this before, probably worth a comment here as to why. Is this to avoid using the nightly compiler?

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/6ebb528c1b64159e6242e897d08784e82381b069..8a945cf94116f92a32ab08191bb0410c298794d6

Split commits

File: twoliter/embedded/Makefile.toml (hunk 18 above)
Comment (webern) [reply twoliter/embedded/Makefile.toml:476]: > The `check-shell` test was previously linting some of the migrated scripts so we should bring that back to this repo at some point.

I have an issue open for this https://github.com/bottlerocket-os/twoliter/issues/42

File: twoliter/build.rs (hunk 20 above)
Comment (webern) [reply twoliter/build.rs:34]: > Do we need to emit `rerun-if-changed` for the non-Rust binary artifacts?

I am! `println!("cargo:rerun-if-changed={}", paths.data_input_dir.display());`

File: twoliter/embedded/Makefile.toml (hunk 19 above)
Comment (webern) [reply twoliter/embedded/Makefile.toml]: > Are we including `infrasys` also? I'd be OK with just dropping it here, not including it as a tool, and removing the sources later.

I moved the sourcecode to here but I am not embedding it in twoliter. I will remove this makefile target.

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/8a945cf94116f92a32ab08191bb0410c298794d6..c6b7e0961a849d2fbd6a330457a9efa2794090f9

Fixup bcressey's suggestions

**Review by bcressey — APPROVED** [review]

**Review by cbgbt — APPROVED** [review]

**Review by bcressey — APPROVED** [review]

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/5406dc0f0187d491c24103f23a6c2f29d2cf3376..14a9919c7609da32e8551c03558104d7c09a608d

Substantial rework of how the tools are embedded. As it turns out, and this makes sense, `build-dependencies` are always built for the host machine. We need the `bindeps` to be `dependencies` so they will be built for the `--target`.

I cannot figure out a way to compress these bytes right now. The `include-flate` macro requires the path to be an actual string literal (not a macro like `env!` that evaluates to a string literal). This might be an oversight (not sure how it all works) that is fixable. But also note that the compilation of `include-flate` seemed to take forever. I'm not sure what is going on there.

File: rust-toolchain.toml [hunk 22]
```diff
@@ -0,0 +1,6 @@
+# This file is locks the build to a specific nightly compiler. There is nothing special about the
```
Comment (cbgbt) [inline rust-toolchain.toml]: nit

```suggestion
# This file locks the build to a specific nightly compiler. There is nothing special about the
```

**Review by cbgbt — APPROVED** [review]

**Review by bcressey — COMMENTED** [review]
Trying to reduce `twoliter` size a bit, I tested [combinations of these settings](https://gist.github.com/bcressey/a9e6d256dc4c6e1a53d13ce58d527ac9) in `Cargo.toml`:
```
[profile.release]
strip = "debuginfo"
codegen-units = 1
lto = true
```

At a bare minimum, I'd recommend stripping debuginfo.

Setting codegen units to 1 is the biggest one but costs the most in terms of added time. If we go there then we may as well go full LTO for the extra 18% size reduction. This seems worthwhile if our release builders can handle it, but definitely not for CI.

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/552f9243c9d7282219fe53e4cd33aab5a821cc6e..9af3a998650d46758e1b34990e39fc23f5f7a7f7

Add release profile with one codegen unit.

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/9af3a998650d46758e1b34990e39fc23f5f7a7f7..84b795a80985ea12524b7d6a75d37e2ac3ef1eb6

Rebase over the latest tools changes from Bottlerocket.

Includes:
- rpm2img xfs fix
- testsys changes

**Review by bcressey — APPROVED** [review]

**Comment by webern** [comment]
> At a bare minimum, I'd recommend stripping debuginfo.

Nice. This change took the binary from 380M to 80M!

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/84b795a80985ea12524b7d6a75d37e2ac3ef1eb6..158e5179fe5e55e39c5e0b505ba598fc8e0aeee7

Fix a typo found by Sean.

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/pull/47/commits/a527e076c885834efd25a22479c714f4cec21c70

Tacked on a commit to stop the unhelpful printing of the entire cargo make invocation which makes it harder for people to find the real error.

**Comment by webern** [comment]
> https://github.com/bottlerocket-os/twoliter/compare/a527e076c885834efd25a22479c714f4cec21c70..ace7b30e6400df5a9af943dfe25b4ebd2cf10536

Rebase to pickup changes from Bottlerocket.
